Progress, BadgeCount, test and fixtures

timborden opened the pull request:
- Updated progress to handle multiple downloads
- Added BadgeCount for number of active downloads

timborden commented:
Before I spotted `app.setBadgeCount` and realized that `win.setOverlayIcon` was only for Windows, I was playing around with an `win.setOverlayIcon` suing HTML5's `<canvas>` element implementation. I don't have a Windows machine, so I can't test it....but here's my WIP (if someone want to pick it up):

```
function setBadge(win, webContents, numDownloadItems) {
	webContents.executeJavaScript(`
		const canvas = document.createElement('canvas');
		const context = canvas.getContext('2d');
		const centerX = canvas.width / 2;
		const centerY = canvas.height / 2;
		const radius = 90;

		context.shadowColor = "grey";
		context.shadowBlur = 10;
		context.shadowOffsetX = 3;
		context.shadowOffsetY = 3;

		context.beginPath();
		context.arc(centerX, centerY, radius, 0, 2 * Math.PI, false);
		context.fillStyle = 'red';
		context.fill();

		context.fillStyle = 'white';
		context.textAlign="center";
		context.textBaseline="middle";
		context.font = '72px sans-serif';
		context.fillText(${numDownloadItems.toString()}, centerX, centerY);
		canvas.toDataURL('image/png', 0.9);
	`)
	.then((result) => {
		const img = nativeImage.createFromDataURL(result);
		win.setOverlayIcon(img, 'Active Downloads');
	});
}
```

sindresorhus commented on index.js:
```diff
@@ -33,19 +39,32 @@ function registerListener(session, opts = {}, cb = () => {}) {
 
 		item.on('updated', () => {
-			const ratio = item.getReceivedBytes() / totalBytes;
+			downloadItems[item.getStartTime()] = {
```
What if two downloads started within the same second?

sindresorhus replied:
I think it would be better to just use a `Set` for `downloadItems`.

https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Set

timborden replied:
Just pushed an update that uses Sets....nice idea.

sindresorhus commented on index.js:
```diff
@@ -33,19 +39,32 @@ function registerListener(session, opts = {}, cb = () => {}) {
 
 		item.on('done', (e, state) => {
-			if (!win.isDestroyed()) {
+			if (['darwin', 'linux'].indexOf(process.platform) >= 0) {
```
Use `Array#includes`

timborden replied:
Good spot....fixed and just pushed

sindresorhus commented:
Can you add a fixture where we can manually test that it works? I don't think there's an easy way to create automated tests for this except for lots of mocking.

Can you also document it in the readme?

timborden commented:
I can update the docs....will do that now.

Not sure about testing. I used `npm start` and manually tested (just hammered my ISP limits....luckily it's the end of the month 😉 ).

timborden commented:
Docs updated....FYI

timborden commented:
Just added the test and fixtures and rebased all the commits into a single commit.

I was able to write a test for the badge count (it was an interesting challenge), but electron doesn't let you getProgress, so there are no tests for that.

Not sure how to resolve the conflicting on my end....the `test.js` has been renamed to run.js and refactored. I've also added a `/test` directory.

Let me know what you think.

kevva commented:
Not sure I'm a big fan of all the test files. I rather test it manually tbh if it can't be shortened in any way.

timborden commented:
@kevva, I'm happy to remove the automated tests and refactor the fixtures for manually testing if that's preferred. What do you think @sindresorhus?

timborden commented:
Hey @sindresorhus & @kevva, any thoughts on this pull request? ...I'm eager to hear your feedback, make the necessary changes and get the commits merged.

Thanks, T

sindresorhus commented:
I'm good with the automated tests. Gives us more confidence in new changes.

sindresorhus commented:
Sorry for the late reply. I haven't been able to do much in open-source the past month because of summer.

sindresorhus commented:
Thank you for working on this @timborden :)

sindresorhus commented:
@timborden Any idea why the CI is now stalling? -> https://travis-ci.org/sindresorhus/electron-dl/builds/261988241

timborden commented:
Hey @sindresorhus,

I'm not sure why it's failing. I can get both base and fork repos tests to pass locally.

I've not used Travis CI before....do think it could be an issue with the output from AVA?
Would it be worth trying the verbose flag `xo && ava -v`?
